scaling-without-state-transfer.asciidoc

```diff
@@ -0,0 +1,97 @@
+= Scaling up without state transfer
+
+The goal is to be able to add nodes to the cluster and make them own new entries,
+without also owning any of the old entries.
+
+Simplifying assumptions:
+
+* Only one node is being added at a time.
+* Single owner
+* No transactions
+* When scaling down data is just lost
+
+The basic idea is that all keys are written on the newest member.
+The location of all inserted keys is kept in a replicated cache
+(the **anchor cache**) and used for further reads/updates/removals.
+
+It's important to note that clients don't have access to the anchor cache,
+so clients access servers in a round-robin fashion.
+
+
+== Implementation
+
+The implementation will live in a module named `anchored-keys`.
+
+Since we can't deterministically map keys to segments, we don't need a distributed cache.
+Instead we can use an invalidation cache and the `ModuleLifecycle` implementation
+can replace the invalidation interceptor with a custom interceptor.
+
+The interceptor will first look up affected keys in the anchor cache.
+If the key doesn't exist in the anchor cache, reads can assume the value is missing.
+Writes must insert the current writer's location in the anchor cache,
+then forward to the current writer for the actual write.
+
+When a key is removed, the interceptor must also remove the key from the anchor cache.
+When a node leaves, the interceptor must remove all the keys mapped to that node
+from the anchor cache.
```

Comment:
I guess that since you don't expect the cache to be reliable (no backups), so here you don't care about concurrency (adding and removing key at the same time might result in anchor cache and the data cache not in sync).
However that also means that you might start leaking entries in the data cache without anchor cache knowing - do you intend to avoid garbage or prune it periodically?

Reply:
No, I do care about concurrency, I just haven't given removals much thought yet.

I'm thinking the interceptor will write a sentinel value in the anchor cache on removal, and a periodic task will prune the sentinel values from the replicated cache. Perhaps setting a lifespan on the sentinel entries would also work.

Writes will have more work to do, but hopefully I can still avoid `compute`, which doesn't yet work with off-heap.